src/java.base/share/classes/java/net/Socket.java

```diff
@@ -518,57 +564,95 @@ private Socket(SocketAddress address, SocketAddress localAddr,
+        if (factory != null) {
+            return factory.createSocketImpl();
+        } else {
+            // create a SOCKS SocketImpl that delegates to a platform SocketImpl
+            SocketImpl delegate = SocketImpl.createPlatformSocketImpl(false);
+            return new SocksSocketImpl(delegate);
         }
     }
 
-    void setImpl(SocketImpl si) {
-         impl = si;
+    /**
+     * Returns the {@code SocketImpl} for this Socket, creating it, and the
+     * underlying socket, if required.
+     * @throws SocketException if creating the underlying socket fails
+     */
+    private SocketImpl getImpl() throws SocketException {
+        if ((state & SOCKET_CREATED) == 0) {
+            synchronized (socketLock) {
+                int s = state;   // re-read state
+                if ((s & SOCKET_CREATED) == 0) {
+                    if (isClosed(s)) {
+                        throw new SocketException("Socket is closed");
+                    }
+                    SocketImpl impl = this.impl;
+                    if (impl == null) {
+                        this.impl = impl = createImpl();
+                    }
+                    try {
+                        impl.create(true);
+                    } catch (SocketException e) {
+                        throw e;
+                    } catch (IOException e) {
+                        throw new SocketException(e.getMessage(), e);
+                    }
+                    getAndBitwiseOrState(SOCKET_CREATED);
+                }
+            }
+        }
+        return impl;
     }
 
     /**
-     * Sets impl to the system-default type of SocketImpl.
-     * @since 1.4
+     * Returns the SocketImpl, may be null.
      */
-    void setImpl() {
-        SocketImplFactory factory = Socket.factory;
-        if (factory != null) {
-            impl = factory.createSocketImpl();
-        } else {
-            // create a SOCKS SocketImpl that delegates to a platform SocketImpl
-            SocketImpl delegate = SocketImpl.createPlatformSocketImpl(false);
-            impl = new SocksSocketImpl(delegate);
+    SocketImpl impl() {
+        return impl;
+    }
+
+    /**
+     * Sets the SocketImpl. The SocketImpl is connected to a peer. The behavior for
+     * the case that the Socket was not a newly created Socket is unspecified. If
+     * there is an existing SocketImpl then it closed to avoid leaking resources.
+     * @throws SocketException if the Socket is closed
+     * @apiNote For ServerSocket use when accepting connections
+     */
+    void setConnectedImpl(SocketImpl si) throws SocketException {
+        synchronized (socketLock) {
+            if ((state & CLOSED) != 0) {
+                throw new SocketException("Socket is closed");
+            }
+            SocketImpl previous = impl;
+            impl = si;
+            state = (SOCKET_CREATED | BOUND | CONNECTED);
+            if (previous != null) {
+                in = null;
+                out = null;
```

Comment:
Hello Alan, I haven't fully grasped this part of the change yet. I see that we now null out these input and outputstreams of the `Socket` instance. If an inputstream was handed out by this `Socket` instance (through a call to `Socket.getInputStream()`) and if the application code calls `InputStream.close()` after this `setConnectedImpl` has been called, then that call to `close()` can lead to closing this newly set `SocketImpl`. Is that OK? Or should that close on the "old" InputStream end up being a no-op? Like I noted, I still haven't fully wrapped my head around this `setConnectedImpl` so maybe this combination of calls is not be possible/practical?

Reply:
ServerSocket.implAccept is intended to be called with a newly created, and unbound, Socket. We looked at this topic in JDK 13 (see JDK-8225431 and related issues) and decided the behavior is unspecified for silly/broken cases where it is called with a Socket in any other state. It should probably have been specified to throw ISE when introduced in JDK 1.1 but we decided there wasn't enough motive to try to fix the spec for that. This PR touches this area again because it interacts with Socket.close.

So the question is what is the JDK behavior when calling implAccept with a connected Socket. A Socket can't have two connections at the same time. The JDK behavior is to attach the SocketImpl for the newly accepted connection to the Socket. This breaks the association from the Socket to the previous SocketImpl. This PR doesn't change that. It does re-visit the question as to what to do with the now orphaned SocketImpl. Should it be discarded, in which case it may be a resource leak, or should it be closed?  The former seems preferable, in which case the read/write methods of streams handed out before calling implAccept will throw IOException. Your question is what happens if the input or output stream close method is called? The spec is that the close method closes the Socket so I've left this as is, it's not changed in this PR. The only real change here is that calling getInputStream/getOutputStream after accepting the new connection will return streams to the newly accepted connection. That seems the most sensible thing to do for this crazy case.

Reply:
Thank you for that detailed explanation, Alan. That helped.

> Should it be discarded, in which case it may be a resource leak, or should it be closed? The former seems preferable, in which case the read/write methods of streams handed out before calling implAccept will throw IOException.

I'm guessing that is a typo and you actually mean "The latter seems preferable ..."? The implementation in this PR  does the latter and closes the previous orphaned SocketImpl and I think that's the right thing to do.

Reply:
Sorry, I meant the former as closing seems preferable,